modules/xml/src/test/java/org/mule/transformers/jaxb/JaxbSecurityTestCase.java

```diff
@@ -0,0 +1,123 @@
+import org.junit.Test;
+
+/**
+ * Test for XML XXE and Billion Laughs attacks
+ */
+public class JaxbSecurityTestCase extends FunctionalTestCase
+{
+
+    @Rule
+    public DynamicPort port = new DynamicPort("port");
+
+    protected final String xmlWithEntities =
+            "<?xml version=\"1.0\"?>\n" +
+            "<!DOCTYPE order [\n" +
+            "<!ELEMENT foo ANY >\n" +
+            "<!ENTITY xxe SYSTEM 'file:%s' >\n" +
+            "<!ENTITY lol \"0101\" >\n" +
+            "]>\n" +
+            "<Foo><bar>&xxe; &lol;</bar></Foo>";
+
+    @Override
+    protected String getConfigFile()
+    {
+        return "jaxb-transformer-security.xml";
+    }
+
+    protected String getXmlWithEntities()
+    {
+        return format(xmlWithEntities, IOUtils.getResourceAsUrl("xxe-passwd.txt", this.getClass()).getPath());
+    }
+
+    protected MuleEvent createEvent() throws Exception
+    {
+        MuleEvent testEvent = getTestEvent(TEST_MESSAGE);
+        testEvent.getMessage().getDataType().setMimeType(APPLICATION_XML);
+        testEvent.getMessage().setPayload(getXmlWithEntities());
+        return testEvent;
+    }
+
+    @Test
+    public void externalEntitiesEnabled() throws Exception
+    {
+        try {
+            System.setProperty(EXTERNAL_ENTITIES_PROPERTY, "true");
+            System.setProperty(EXPAND_ENTITIES_PROPERTY, "true");
+
+            final MuleEvent testEvent = createEvent();
+            final MuleEvent response = ((Flow) getFlowConstruct("testFlow")).process(testEvent);
+
+            String payload = response.getMessage().getPayloadAsString();
+
+            assertThat(payload, containsString("secret"));
+            assertThat(payload, containsString("0101"));
+        }
+        finally
+        {
+            System.clearProperty(EXTERNAL_ENTITIES_PROPERTY);
+            System.clearProperty(EXPAND_ENTITIES_PROPERTY);
+        }
+    }
+
+    @Test
+    public void expandsEntitiesEnabled() throws Exception
+    {
+        try {
+            System.setProperty(EXPAND_ENTITIES_PROPERTY, "true");
+
+            final MuleEvent testEvent = createEvent();
+            final MuleEvent response = ((Flow) getFlowConstruct("testFlow")).process(testEvent);
+
+            String payload = response.getMessage().getPayloadAsString();
+
+            assertThat(payload, not(containsString("secret")));
+            assertThat(payload, containsString("0101"));
+        }
+        finally
+        {
+            System.clearProperty(EXPAND_ENTITIES_PROPERTY);
+        }
+    }
+
+    @Test
+    public void expandsEntitiesWhenDisabled() throws Exception
+    {
+        try
+        {
+            final MuleEvent testEvent = createEvent();
+            final MuleEvent response = ((Flow) getFlowConstruct("testFlow")).process(testEvent);
+
+            String payload = response.getMessage().getPayloadAsString();
+
+            fail("Should've thrown exception");
+        } catch (TransformerMessagingException e) {
```

Comment:
We use ExpectedException instead of try/catch, see examples from other junits and change this is possible.